src/blob/index/bptree/core.rs

```diff
@@ -0,0 +1,359 @@
+/// structure of b+-tree index file from the beginning:
+/// 1. Header
+/// 2. b+-tree user buffer (now Bloom filter is stored as this buffer)
+/// 3. Tree metadata (tree_offset [4] (tree starts from root, so that is also root_offset),
+///    and leaves_offset [5])
+/// 4. Tree: Non-leaf nodes of bptree in format
+///    `NodeMeta | keys_arr | pointers_arr`,
+///    where pointer is offset in file of underlying node with searched key
+/// 5. Sorted by key array of record headers (coupled leaf nodes)
+use super::prelude::*;
+
+pub(super) const BLOCK_SIZE: usize = 4096;
+
+#[derive(Debug, Clone)]
+pub struct BPTreeFileIndex {
+    file: File,
+    header: IndexHeader,
+    metadata: TreeMeta,
+    root_node: [u8; BLOCK_SIZE],
+}
+
+#[async_trait::async_trait]
+impl FileIndexTrait for BPTreeFileIndex {
+    async fn from_file(name: FileName, ioring: Option<Rio>) -> Result<Self> {
+        trace!("open index file");
+        let file = File::open(name.to_path(), ioring)
+            .await
+            .context(format!("failed to open index file: {}", name))?;
+        let header = Self::read_index_header(&file).await?;
+        let metadata = Self::read_tree_meta(&file, &header).await?;
+        let root_node = Self::read_root(&file, metadata.tree_offset).await?;
+
+        Ok(Self {
+            file,
+            header,
+            metadata,
+            root_node,
+        })
+    }
+
+    async fn from_records(
+        path: &Path,
+        ioring: Option<Rio>,
+        headers: &InMemoryIndex,
+        meta: Vec<u8>,
+        recreate_index_file: bool,
+    ) -> Result<Self> {
+        clean_file(path, recreate_index_file)?;
+        let res = Self::serialize(headers, meta)?;
+        let (mut header, metadata, buf) = res;
+        let file = File::create(path, ioring)
+            .await
+            .with_context(|| format!("file open failed {:?}", path))?;
+        file.write_append(&buf).await?;
+        header.set_written(true);
+        let serialized_header = serialize(&header)?;
+        file.write_at(0, &serialized_header).await?;
+        file.fsyncdata().await?;
+        let root_node = Self::read_root(&file, metadata.tree_offset).await?;
+        Ok(Self {
+            file,
+            metadata,
+            header,
+            root_node,
+        })
+    }
+
+    fn file_size(&self) -> u64 {
```

Comment:
I think this method is redundant, because it is not public so you already have access to file field